Assets/Scripts/Scheduler/Scheduler.cs

```diff
@@ -0,0 +1,241 @@
+
+        /// <summary>
+        /// Schedules an event from a prototype.
+        /// </summary>
+        /// <param name="name">Name of the event (for serialization etc.).</param>
+        /// <param name="cooldown">Cooldown in seconds.</param>
+        /// <param name="repeatsForever">Whether the event repeats forever (defaults false). If true repeats is ignored.</param>
+        /// <param name="repeats">Number of repeats (default 1). Ignored if repeatsForever=true.</param>
+        public void ScheduleEvent(string name, float cooldown, bool repeatsForever = false, int repeats = 1)
+        {
+            if (EventPrototypes.ContainsKey(name) == false)
+            {
+                Debug.ULogWarningChannel("Scheduler", "Tried to schedule an event from a prototype '{0}' which does not exist. Bailing.", name);
+                return;
+            }
+
+            EventPrototype ep = EventPrototypes[name];
+            ScheduledEvent evt = new ScheduledEvent(ep, cooldown, cooldown, repeatsForever, repeats);
+
+            RegisterEvent(evt);
+        }
+
+        public void RegisterEvent(ScheduledEvent evt)
+        {
+            if (evt != null)
+            {
+                if (IsRegistered(evt))
+                {
+                    Debug.ULogChannel("Scheduler", "Event '{0}' registered more than once.", evt.Name);
+                }
+
+                Events.Add(evt);
+            }
+        }
+
+        public bool IsRegistered(ScheduledEvent evt)
+        {
+            return Events != null && Events.Contains(evt);
+        }
+
+        /// <summary>
+        /// Deregisters the event.
+        /// NOTE: To stop a continuing event from running do not try to deregister it with this!
+        /// Instead call Stop() on the event. Then it will be removed on the next purge.
+        /// </summary>
+        /// <param name="evt">Event to deregister.</param>
+        public void DeregisterEvent(ScheduledEvent evt)
+        {
+            if (Events != null)
+            {
+                Events.Remove(evt);
+            }
+        }
+
+        public void PurgeEventList()
+        {
+            if (Events != null)
+            {
+                Events.RemoveAll((e) => e.Finished);
+            }
+        }
+
+        /// <summary>
+        /// Time to next event in seconds.
+        /// </summary>
+        /// <returns>The to next event (-1 if there are no events).</returns>
+        public float TimeToNextEvent()
+        {
+            if (Events == null || Events.Count == 0)
+            {
+                return -1f;
+            }
+
+            return Events.Min((e) => e.TimeToWait);
+        }
+
+        public void Update(float deltaTime)
+        {
+            if (Events == null || Events.Count == 0)
+            {
+                return;
+            }
+
+            // Events may try to modify the event list, so cannot use a foreach on Events!
+            ScheduledEvent[] eventArray = Events.ToArray();
```

Comment:
Instead of modifying the Event list directly in `RegisterEvent()` and `DeregisterEvent()`. You could queue up the changes to the list, then apply them here, before the loop. This way you can avoid coping the Event list on every tick.